sharktank/sharktank/tools/tuner/candidate_gen.py

```diff
@@ -0,0 +1,1386 @@
+
+    return constraints
+
+
+def generate_solutions(problem_size: ProblemSize, num_subgrups: int):
+    M, N, K = problem_size.MNK
+    tune_logger.info(f"{M},{N},{K}")
+    m, n, k = z3.Int("m"), z3.Int("n"), z3.Int("k")
+    subgroup_size = z3.Int("subgroup_size")
+    intrinsic_mn = z3.Int("intrinsic_mn")
+    intrinsic_k = z3.Int("intrinsic_k")
+    wg_x, wg_y, wg_z = z3.Int("wg_x"), z3.Int("wg_y"), z3.Int("wg_z")
+    sg_m_cnt = z3.Int("sg_m_cnt")
+    sg_n_cnt = z3.Int("sg_n_cnt")
+    waves_per_eu = z3.Int("waves_per_eu")
+    all_vars = [
+        m,
+        n,
+        k,
+        subgroup_size,
+        intrinsic_mn,
+        intrinsic_k,
+        wg_x,
+        wg_y,
+        wg_z,
+        sg_m_cnt,
+        sg_n_cnt,
+        waves_per_eu,
+    ]
+
+    solver = z3.Solver()
+    constraints = generate_constraints(
+        problem_size,
+        [m, n, k],
+        num_subgrups,
+        subgroup_size,
+        [intrinsic_mn, intrinsic_k],
+        [wg_x, wg_y, wg_z],
+        sg_m_cnt,
+        sg_n_cnt,
+        waves_per_eu,
+    )
+    solver.add(z3.simplify(z3.And(constraints)))
+    tune_logger.debug(f"Initial constraints: {solver}")
+    i = 0
+    while solver.check() == z3.sat:
+        model = solver.model()
+        lookup = lambda var: model[var].as_long()
+
+        config = Configuration(
+            lookup(subgroup_size),
+            [lookup(wg_x), lookup(wg_y), lookup(wg_z)],
+            MfmaIntrinsic(
+                problem_size.lhs_type.element_type,
+                lookup(intrinsic_mn),
+                lookup(intrinsic_mn),
+                lookup(intrinsic_k),
+                problem_size.res_type.element_type,
+            ),
+            [lookup(m), lookup(n), lookup(k)],
+            lookup(sg_m_cnt),
+            lookup(sg_n_cnt),
+            lookup(waves_per_eu),
+        )
+        solver.add(z3.simplify(z3.Not(z3.And(list(x == model[x] for x in all_vars)))))
+        i += 1
+        yield config
+
+
+def get_default_output_dir() -> str:
+    from datetime import datetime
+
+    return "tuning_" + datetime.now().strftime("%Y_%m_%d_%H_%M")
+
+
+def parse_mlir(mlir_text: str) -> ir.Module:
+    mlir_module = None
+    with ireec.ir.Context() as context:
+        try:
+            mlir_module = ireec.ir.Module.parse(mlir_text)
+            tune_logger.info("MLIR parsing successful!")
+        except ireec.ir.MLIRError as e:
+            tune_logger.error(f"Error parsing MLIR: {e}")
+            raise RuntimeError(f"Error parsing MLIR: {e}")
+
+    return mlir_module
+
+
+@dataclass
+class CandidateGenFn:
+    get_shapes_fn: Optional[Callable[[list[str]], ProblemSize]] = None
+    apply_params_fn: Optional[
+        Callable[[ProblemSize, list[str], Configuration], tuple[str, str]]
+    ] = None
+
+
+class DispatchTuner(ABC):
+    @abstractmethod
+    def supports(self, mlir: str) -> bool:
+        pass
+
+    @abstractmethod
+    def get_candidate_gen_fn(self) -> CandidateGenFn:
+        pass
```

Comment:
Nice! I think we could go one step further and instead of `DispatchTuner` returning `CandidateGenFn`, we can make these two functions be methods in `DispatchTuner`, and convert all these helper functions `get_shapes_*`, `apply_params_*` , etc. to methods in each tuner subclass.